geth-utils/gethutil/mpt/trie/stacktrie.go

```diff
@@ -551,50 +551,49 @@ func (st *StackTrie) Commit() (common.Hash, error) {
 	return common.BytesToHash(st.val), nil
 }
 
-func (st *StackTrie) getNodeFromBranchRLP(branch []byte, ind byte) []byte {
-	start := 2 // when branch[0] == 248
-	if branch[0] == 249 {
-		start = 3
-	}
-
-	i := 0
-	insideInd := -1
-	cInd := byte(0)
-	for {
-		if start+i == len(branch)-1 { // -1 because of the last 128 (branch value)
-			return []byte{0}
-		}
-		b := branch[start+i]
-		if insideInd == -1 && b == 128 {
-			if cInd == ind {
+const RLP_SHORT_STR_FLAG = 128
+const RLP_LONG_LIST_FLAG = 248
+const LEN_OF_HASH = 32
+
+// Note:
+// In RLP encoding, if the value is between [0x80, 0xb7] ([128, 183]),
+// it means following data is a short string (0 - 55bytes).
+// Which implies if the value is 128, it's an empty string.
+func (st *StackTrie) getNodeFromBranchRLP(branch []byte, idx int) []byte {
+	// (branch[0] - RLP_LONG_LIST_FLAG) means the length in bytes of the length of the payload
+	// and the payload is right after the length.
+	// That's why we add `2` here
+	// e.g. [248 81 128 160 ...]
+	// `81` is the length of the payload and payload starts from `128`
+	start := int(branch[0]) - RLP_LONG_LIST_FLAG + 2
+
+	// If 1st node is not 128(empty node) or 160, it should be a leaf
+	b := int(branch[start])
+	if b != RLP_SHORT_STR_FLAG || b != (RLP_SHORT_STR_FLAG+LEN_OF_HASH) {
+		return []byte{0}
+	}
+
+	current_idx := 0
+	for i := start; i < len(branch); i++ {
+		b = int(branch[i])
+		switch b {
+		case RLP_SHORT_STR_FLAG: // 128
+			// if the current index is we're looking for, return an empty node directly
+			if current_idx == idx {
 				return []byte{128}
-			} else {
-				cInd += 1
 			}
-		} else if insideInd == -1 && b != 128 {
-			if b == 160 {
-				if cInd == ind {
-					return branch[start+i+1 : start+i+1+32]
-				}
-				insideInd = 32
-			} else {
-				// non-hashed node
-				if cInd == ind {
-					return branch[start+i+1 : start+i+1+int(b)-192]
-				}
-				insideInd = int(b) - 192
-			}
-			cInd += 1
-		} else {
-			if insideInd == 1 {
-				insideInd = -1
-			} else {
-				insideInd--
+			current_idx++
+		case RLP_SHORT_STR_FLAG + LEN_OF_HASH: // 160
```

Comment:
Note that `b = 160` could also be for any of the bytes of the branch child, it's not always the beginning of the new child, you need to somehow remember the length of the child and check whether you exhausted all the bytes of the child.

Regarding the child length - I think here you can rely it's either 32 or 1, because all the children are hashed. The child is not hashed if it's shorter than 32 bytes, but for transactions, receipts, and withdrawals I think this is never the case.

Reply:
Yep, I know `160` is not always the beginning of a new child, could be any of the bytes in a branch. My refactor is trying to parse RLP data instead of going through every single byte. And I found for transactions case, there are only `128` and `160` so I remove the original logic of handling `192`.  If I only focus on the case for transactions, wondering if this implementation make sense to you?

>Regarding the child length - I think here you can rely it's either 32 or 1, because all the children are hashed. The child is not hashed if it's shorter than 32 bytes, but for transactions, receipts, and withdrawals I think this is never the case.

I'm not sure I follow this. Do you mean for the case of transactions, it's always 1 (or 0, I'm assuming you're talking about an empty node) or 32 bytes? But above cases doesn't apply to receipts and withdraws, right? 

Could you elebarate more one this "The child is not hashed  if it's shorter than 32 bytes". Why is it shorter than 32 bytes if the child is not hashed? Won't it longer than 32 bytes?

Reply:
But doesn't this code switch to the new child as soon as 128 or 160 is encountered (even if it's in the middle of the child and this byte actually means the child byte and not RLP encoding)?

The idea is to use this function for receipts and withdrawals as well.

When the leaf is shorter than 32 bytes, it won't get hashed when put in the branch RLP, it will be put in the branch as it is and its RLP encoding byte won't be 160 (its length won't be 32).

Reply:
I didn't make it clear. Yes, this code switch to a new child as soon as 128 or 160. But my assumption is it won't jump into the middle of the child because it starts from index 2 (or 3, depending on starting byte is 248 or 249) and then it moves to next byte if it's a 128 or moves 32 bytes if it's a 160. If there are only 128 or 160 in the branch, then it should be fine. Is it too naive?

Agree, we should use this function for receipts ans withdrawals. However, I want to move a small step at a time and make sure everything works for transactions. That's why I only take care the case of transactions. If you think it's better to cover receipts and withdrawals, I could take time to dive into their format. That works for me as well!

I didn't realize the leaf will be shorter than 32 bytes here. I'll figure it out how to fix it.

Reply:
Yes, sorry, didn't realise you are increasing `i` by the length of the child. It should work fine. For non-hashed nodes I think not many changes are needed, you need to compute the length of the node as `b - 192`.

Agree on the incremental approach, we can take care of receipts and withdrawals later.